iree/samples/dynamic_shapes/CMakeLists.txt

```diff
@@ -4,9 +4,7 @@
 # See https://llvm.org/LICENSE.txt for license information.
 # SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 
-# iree_runtime_runtime -> iree::hal::drivers dependency requires threading
-# TODO(#4298): Separate HAL driver registration from threading.
-if(NOT ${IREE_ENABLE_THREADING})
+if(NOT ${IREE_HAL_DRIVER_DYLIB})
   return()
 endif()
```

Comment:
This sample has no specific dependency on a HAL driver. If the threading issue is fixed it should have no condition here.

Reply:
`test.sh` calls for dylib driver, so I think we need to check to make sure it is built in `iree/hal/drivers`

Reply:
Ahh... sorry I only read the last part of the script. Yes the condition should be removed.

Reply:
Actually, we need to keep this condition check to guarded from the bare-metal build.

Reply:
How so?

Reply:
```
[2/2] Linking C executable iree/samples/dynamic_shapes/dynamic-shapes
FAILED: iree/samples/dynamic_shapes/dynamic-shapes 
: && .../toolchain_iree_rv32imf/bin/clang -march=rv32imf -mabi=ilp32 -DIREE_PLATFORM_GENERIC=1 
-DIREE_SYNCHRONIZATION_DISABLE_UNSAFE=1       -DIREE_FILE_IO_ENABLE=0 -DIREE_TIME_NOW_FN="{
return 0; }" -DIREE_DEVICE_SIZE_T=uint32_t -DPRIdsz=PRIu32   -O3 -DNDEBUG -lm      -lm 
iree/samples/dynamic_shapes/CMakeFiles/iree_samples_dynamic_shapes.dir/main.c.obj -o 
iree/samples/dynamic_shapes/dynamic-shapes  iree/runtime/libiree_runtime_impl.a  
iree/base/internal/libiree_base_internal_file_io.a  iree/hal/drivers/libiree_hal_drivers_drivers.a 
iree/hal/dylib/registration/libiree_hal_dylib_registration_sync.a  
iree/hal/local/loaders/libiree_hal_local_loaders_embedded_library_loader.a  
iree/hal/local/elf/libiree_hal_local_elf_elf_module.a  iree/hal/local/elf/libiree_hal_local_elf_arch.a 
iree/hal/local/elf/libiree_hal_local_elf_platform.a  iree/hal/local/loaders/libiree_hal_local_loaders_system_library_loader.a  
iree/base/internal/libiree_base_internal_dynamic_library.a  -ldl  iree/base/internal/libiree_base_internal_file_path.a  
iree/hal/vmvx/registration/libiree_hal_vmvx_registration_sync.a  iree/hal/local/libiree_hal_local_sync_driver.a 
iree/base/internal/libiree_base_internal_arena.a  iree/base/internal/libiree_base_internal_atomic_slist.a  
iree/hal/local/loaders/libiree_hal_local_loaders_vmvx_module_loader.a  iree/hal/local/libiree_hal_local_local.a  
iree/modules/vmvx/libiree_modules_vmvx_vmvx.a  iree/modules/hal/libiree_modules_hal_hal.a  iree/hal/libiree_hal_hal.a  
iree/base/internal/libiree_base_internal_synchronization.a  iree/vm/libiree_vm_bytecode_module.a  
iree/vm/libiree_vm_impl.a  build_tools/third_party/flatcc/libflatcc_parsing.a  -lm  iree/base/libiree_base_base.a && :
.../toolchain_iree_rv32imf/bin/riscv32-unknown-elf-ld: cannot find -ldl
```

Reply:
These samples _only_ depend on `iree_runtime_runtime`. Isn't one of the goals of this PR to make that not always depend on threading / OS libraries? (replacing the threading check with another cmake check would be okay, but this should not depend on a specific HAL driver)

Reply:
This PR will not completely fix #4298, but only separate the hard dependency of driver registration from threading, so we can enable iree/runtime for the threadless configuration. I will keep the issue note in the CMakeLists.txt so we know it still needs to be cleaned later.

Reply:
The latest commit (make `dylib-sync` driver compatible with only embedded library loader) should fix the driver-flag issue.